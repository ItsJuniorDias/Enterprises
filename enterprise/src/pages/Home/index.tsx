/* eslint-disable no-bitwise */
import React, { useState, useEffect, useRef } from 'react';
import { View, Image, FlatList, Text } from 'react-native';
import { useDispatch, useSelector } from 'react-redux';

import { useNavigation } from '@react-navigation/native';

import { IShow } from 'store/modules/show/types';
import exit from '../../assets/Vector.png';
import not from '../../assets/not.png';

import { logoutUser } from '../../store/modules/auth/actions';
import { IEnterprise } from '../../store/modules/enterprise/types';

import { requestShow } from '../../store/modules/show/actions';

import InputSearch from '../../components/InputSearch';
import Loading from '../../components/Loading';

import api from '../../services/api';

import {
  Container,
  Description,
  Header,
  Title,
  TouchableExit,
  ContentFlat,
  ContainerCard,
  ImageCard,
  ShadowContent,
  TitleCard,
  DescriptionCard,
  ContentDescription,
  ViewShimmer,
  EmptyFilterData,
  ContentEmpty,
} from './styles';

const Home: React.FC = () => {
  const [dataEnterprise, setDataEnterprise] = useState<IEnterprise[]>([]);

  const [loading, setLoading] = useState(false);
  const [filter, setFilter] = useState(false);

  const [dataFiltered, setDataFiltered] = useState<IEnterprise[]>([]);

  console.log(dataFiltered, 'FILTERED');

  const enterprise = useSelector(state => state.enterprise);

  const headers = useSelector(state => state.show);

  const dispatch = useDispatch();

  const navigation = useNavigation();

  useEffect(() => {
    setDataEnterprise(Object.values(enterprise));
  }, [enterprise]);

  const handeExit = () => {
    dispatch(logoutUser());

    navigation.navigate('/SignIn');
  };

  const handleShow = async (id: IShow) => {
    dispatch(requestShow(headers, id));

    navigation.navigate('/Show');
  };

  const CardItem = ({ title, description, photo, id }) => {
    return (
      <>
        <ContainerCard onPress={() => handleShow(id)}>
          <ImageCard
            source={{
              uri: `https://empresas.ioasys.com.br/${photo}`,
            }}
          />
          <ShadowContent>
            <TitleCard>{title}</TitleCard>

            <ContentDescription>
              <DescriptionCard numberOfLines={4}>{description}</DescriptionCard>
            </ContentDescription>
          </ShadowContent>
        </ContainerCard>
      </>
    );
  };

  const renderItem = ({ item }) => (
    <CardItem
      key={item.id}
      id={item.id}
      title={item.enterprise_name}
      description={item.description}
      photo={item.photo}
    />
  );

  const handleFilterValue = (value: any) => {
    const itemFiltered = dataEnterprise.filter(
      item =>
        item.enterprise_name.toLowerCase().includes(value.toLowerCase()) ||
        item.description.toLowerCase().includes(value.toLowerCase()),
    );

    if (itemFiltered) {
      setDataFiltered(itemFiltered);
      setFilter(true);
    } else {
      setFilter(false);
    }
  };

  const emptyListDataFilter = () => (
    <>
      <ContentEmpty>
        <EmptyFilterData> Não Encontrado</EmptyFilterData>
        <Image source={not} />
      </ContentEmpty>
    </>
  );

  return (
    <>
      {loading && <Loading />}
      {!loading && (
        <>
          <Container>
            <Header>
              <View>
                <Title>Olá, Alexandre!</Title>
                <Description>Bem-vindo(a)</Description>
              </View>

              <TouchableExit onPress={() => handeExit()}>
                <Image source={exit} />
              </TouchableExit>
            </Header>

            <InputSearch
              title="Buscar por nome"
              name="filter"
              callBackParent={value => handleFilterValue(value)}
            />

            <ContentFlat>
              {filter ? (
                <FlatList
                  data={dataFiltered}
                  ListEmptyComponent={emptyListDataFilter}
                  renderItem={renderItem}
                  keyExtractor={(item, index) => {
                    return index.toString();
                  }}
                />
              ) : (
                <FlatList
                  data={dataEnterprise}
                  renderItem={renderItem}
                  keyExtractor={(item, index) => {
                    return index.toString();
                  }}
                />
              )}
            </ContentFlat>
          </Container>
        </>
      )}
    </>
  );
};

export default Home;
