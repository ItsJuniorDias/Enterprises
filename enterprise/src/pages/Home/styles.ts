import styled from 'styled-components/native';

export const Container = styled.View`
  flex: 1;
  padding: 0 30px;
`;

export const Header = styled.View`
  width: 100%;
  height: 140px;

  flex-direction: row;
  align-items: center;
  justify-content: space-between;
`;

export const Title = styled.Text`
  font-size: 18px;
  font-family: 'Poppins-Regular';
  color: #2d4379;
  line-height: 24px;
  margin-bottom: 8px;
`;

export const Description = styled.Text`
  font-family: 'Poppins-SemiBold';

  font-weight: 800;
  font-size: 24px;
  line-height: 33px;

  color: #0d253c;
`;

export const TouchableExit = styled.TouchableOpacity`
  margin-top: -32px;
  width: 48px;
  height: 48px;

  border-radius: 50px;
  align-items: center;
  justify-content: center;
`;

export const ContentFlat = styled.SafeAreaView`
  flex: 1;
  margin-top: 24px;
`;

export const ContainerCard = styled.TouchableOpacity`
  width: 100%;
  height: 162px;

  margin-bottom: 32px;
  border-radius: 8px;
`;

export const ImageCard = styled.Image`
  width: 100%;
  height: 162px;
  position: absolute;
  z-index: 2;
  border-radius: 8px;
`;

export const ShadowContent = styled.View`
  width: 100%;
  height: 162px;
  position: absolute;
  z-index: 3;
  border-radius: 8px;
  padding: 16px;
  background-color: rgba(0, 0, 0, 0.4);
`;

export const TitleCard = styled.Text`
  font-family: 'Poppins-SemiBold';

  font-size: 18px;
  z-index: 5;

  /* identical to box height */

  color: #ffffff;
`;

export const ContentDescription = styled.View`
  height: 80px;
`;

export const DescriptionCard = styled.Text`
  font-family: 'Poppins-SemiBold';

  font-size: 14px;
  z-index: 5;

  /* identical to box height */

  color: #ffffff;
`;
