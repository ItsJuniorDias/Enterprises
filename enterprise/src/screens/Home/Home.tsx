import React, { useState, useEffect } from 'react';
import { View, Image, FlatList } from 'react-native';
import { useDispatch, useSelector } from 'react-redux';
import { useNavigation } from '@react-navigation/native';
import { IShow } from 'store/modules/show/types';
import exit from '../../assets/Vector.png';
import not from '../../assets/not.png';
import { logoutUser } from '../../store/modules/auth/actions';
import { IEnterprise } from '../../store/modules/enterprise/types';
import { requestShow } from '../../store/modules/show/actions';
import { InputSearch } from '../../components';

import { useAuth } from 'hooks';

import {
  Container,
  Description,
  Header,
  Title,
  TouchableExit,
  ContentFlat,
  ContainerCard,
  ImageCard,
  ShadowContent,
  TitleCard,
  DescriptionCard,
  ContentDescription,
  EmptyFilterData,
  ContentEmpty,
} from './styles';

export const Home = () => {
  const [dataEnterprise, setDataEnterprise] = useState<IEnterprise[]>([]);
  const [filter, setFilter] = useState(false);
  const [dataFiltered, setDataFiltered] = useState<IEnterprise[]>([]);

  const enterprise = useSelector((state) => state.enterprise);
  const headers = useSelector((state) => state.show);

  const dispatch = useDispatch();
  const navigation = useNavigation();

  const { logout } = useAuth();

  useEffect(() => {
    setDataEnterprise(Object.values(enterprise));
  }, [enterprise]);

  const handeExit = () => {
    logout();

    navigation.navigate('/SignIn');
  };

  const handleShow = async (id: IShow) => {
    dispatch(requestShow(headers, id));
    navigation.navigate('/Show');
  };

  const CardItem = ({ title, description, photo, id }) => {
    return (
      <>
        <ContainerCard
          testID="containerCard_testId"
          onPress={() => handleShow(id)}
        >
          <ImageCard
            source={{
              uri: `https://empresas.ioasys.com.br/${photo}`,
            }}
          />
          <ShadowContent>
            <TitleCard>{title}</TitleCard>

            <ContentDescription>
              <DescriptionCard numberOfLines={4}>{description}</DescriptionCard>
            </ContentDescription>
          </ShadowContent>
        </ContainerCard>
      </>
    );
  };

  const renderItem = ({ item }) => (
    <CardItem
      key={item.id}
      id={item.id}
      title={item.enterprise_name}
      description={item.description}
      photo={item.photo}
    />
  );

  const handleFilterValue = (value: any) => {
    const itemFiltered = dataEnterprise.filter(
      (item) =>
        item.enterprise_name.toLowerCase().includes(value.toLowerCase()) ||
        item.description.toLowerCase().includes(value.toLowerCase())
    );

    if (!!itemFiltered.length) {
      setDataFiltered(itemFiltered);
      setFilter(true);
    } else {
      setFilter(false);
    }
  };

  const emptyListDataFilter = () => (
    <>
      <ContentEmpty>
        <EmptyFilterData> Não Encontrado</EmptyFilterData>
        <Image source={not} />
      </ContentEmpty>
    </>
  );

  return (
    <>
      <Container>
        <Header>
          <View>
            <Title>Olá, Alexandre!</Title>
            <Description>Bem-vindo(a)</Description>
          </View>

          <TouchableExit testID="buttonExit_testId" onPress={() => handeExit()}>
            <Image source={exit} />
          </TouchableExit>
        </Header>

        <InputSearch
          title="Buscar por nome"
          name="filter"
          callBackParent={(value) => handleFilterValue(value)}
        />

        <ContentFlat>
          {filter ? (
            <FlatList
              data={dataFiltered}
              ListEmptyComponent={emptyListDataFilter}
              renderItem={renderItem}
              keyExtractor={(_, index) => {
                return index.toString();
              }}
            />
          ) : (
            <FlatList
              data={dataEnterprise}
              renderItem={renderItem}
              keyExtractor={(_, index) => {
                return index.toString();
              }}
            />
          )}
        </ContentFlat>
      </Container>
    </>
  );
};
