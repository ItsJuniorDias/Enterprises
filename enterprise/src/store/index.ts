import { createStore, applyMiddleware, compose } from 'redux';
import { composeWithDevTools } from 'redux-devtools-extension';
import createSagaMiddleware from 'redux-saga';

import rootReducer from './modules/rootReducer';

import { IAuthState } from './modules/auth/types';
import rootSaga from './modules/rootSaga';

export interface IState {
  auth: IAuthState;
}

const sagaMiddleware = createSagaMiddleware();
const middlewares = [sagaMiddleware];

const store = createStore(
  rootReducer,
  composeWithDevTools(applyMiddleware(...middlewares)),
);

sagaMiddleware.run(rootSaga);

export default store;
