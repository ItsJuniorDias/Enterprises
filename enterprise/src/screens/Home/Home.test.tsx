import React from 'react';
import { render } from '@testing-library/react-native';
import { Home } from '../Home/Home';
import { Provider } from 'react-redux';
import store from 'store';

jest.mock('react-native-shimmer-placeholder', () => {
  return {
    createShimmerPlaceholder: jest.fn(),
  };
});
jest.mock('react-native-linear-gradient', () => {});
jest.mock('@react-navigation/native', () => ({
  useNavigation: jest.fn(),
}));

describe('Behavior Home', () => {
  const screenRender = (
    <Provider store={store}>
      <Home />
    </Provider>
  );

  it('render snapshot', () => {
    const result = render(screenRender).toJSON();

    expect(result).toMatchSnapshot();
  });
});
