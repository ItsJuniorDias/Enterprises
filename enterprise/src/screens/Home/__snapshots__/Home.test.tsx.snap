// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Behavior Home render snapshot 1`] = `
<View
  style={
    Array [
      Object {
        "flexBasis": 0,
        "flexGrow": 1,
        "flexShrink": 1,
        "paddingBottom": 0,
        "paddingLeft": 30,
        "paddingRight": 30,
        "paddingTop": 0,
      },
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "height": 140,
          "justifyContent": "space-between",
          "width": "100%",
        },
      ]
    }
  >
    <View>
      <Text
        style={
          Array [
            Object {
              "color": "#2d4379",
              "fontFamily": "Poppins-Regular",
              "fontSize": 18,
              "lineHeight": 24,
              "marginBottom": 8,
            },
          ]
        }
      >
        Ol√°, Alexandre!
      </Text>
      <Text
        style={
          Array [
            Object {
              "color": "#0d253c",
              "fontFamily": "Poppins-SemiBold",
              "fontSize": 24,
              "fontWeight": "800",
              "lineHeight": 33,
            },
          ]
        }
      >
        Bem-vindo(a)
      </Text>
    </View>
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "borderRadius": 50,
          "height": 48,
          "justifyContent": "center",
          "marginTop": -32,
          "opacity": 1,
          "width": 48,
        }
      }
    >
      <Image
        source={
          Object {
            "testUri": "../../../src/assets/Vector.png",
          }
        }
      />
    </View>
  </View>
  <View
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "borderColor": "#f3f3f3",
        "borderRadius": 25,
        "borderWidth": 1,
        "flexDirection": "row",
        "height": 50,
        "marginTop": -8,
        "opacity": 1,
        "paddingBottom": 0,
        "paddingLeft": 16,
        "paddingRight": 16,
        "paddingTop": 0,
        "width": "100%",
      }
    }
  >
    <Image
      source={
        Object {
          "testUri": "../../../src/assets/search.png",
        }
      }
    />
    <View
      style={
        Array [
          Object {
            "width": "90%",
          },
        ]
      }
    >
      <TextInput
        allowFontScaling={true}
        onChangeText={[Function]}
        placeholder="Buscar por nome"
        placeholderTextColor="#A4A4B2"
        rejectResponderTermination={true}
        style={
          Array [
            Object {
              "color": "#a4a4b2",
              "fontFamily": "Poppins-Regular",
              "marginLeft": 8,
              "marginTop": 4,
            },
          ]
        }
        underlineColorAndroid="transparent"
      />
    </View>
  </View>
  <RCTSafeAreaView
    emulateUnlessSupported={true}
    style={
      Array [
        Object {
          "flexBasis": 0,
          "flexGrow": 1,
          "flexShrink": 1,
          "marginTop": 24,
        },
      ]
    }
  >
    <RCTScrollView
      data={
        Array [
          Object {},
        ]
      }
      disableVirtualization={false}
      getItem={[Function]}
      getItemCount={[Function]}
      horizontal={false}
      initialNumToRender={10}
      keyExtractor={[Function]}
      maxToRenderPerBatch={10}
      onContentSizeChange={[Function]}
      onEndReachedThreshold={2}
      onLayout={[Function]}
      onMomentumScrollEnd={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      removeClippedSubviews={false}
      renderItem={[Function]}
      scrollEventThrottle={50}
      stickyHeaderIndices={Array []}
      updateCellsBatchingPeriod={50}
      viewabilityConfigCallbackPairs={Array []}
      windowSize={21}
    >
      <View>
        <View
          onLayout={[Function]}
          style={null}
        >
          <View
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "borderRadius": 8,
                "height": 162,
                "marginBottom": 32,
                "opacity": 1,
                "width": "100%",
              }
            }
          >
            <Image
              source={
                Object {
                  "uri": "https://empresas.ioasys.com.br/undefined",
                }
              }
              style={
                Array [
                  Object {
                    "borderRadius": 8,
                    "height": 162,
                    "position": "absolute",
                    "width": "100%",
                    "zIndex": 2,
                  },
                ]
              }
            />
            <View
              style={
                Array [
                  Object {
                    "backgroundColor": "rgba(0, 0, 0, 0.4)",
                    "borderRadius": 8,
                    "height": 162,
                    "paddingBottom": 16,
                    "paddingLeft": 16,
                    "paddingRight": 16,
                    "paddingTop": 16,
                    "position": "absolute",
                    "width": "100%",
                    "zIndex": 3,
                  },
                ]
              }
            >
              <Text
                style={
                  Array [
                    Object {
                      "color": "#ffffff",
                      "fontFamily": "Poppins-SemiBold",
                      "fontSize": 18,
                      "zIndex": 5,
                    },
                  ]
                }
              />
              <View
                style={
                  Array [
                    Object {
                      "height": 80,
                    },
                  ]
                }
              >
                <Text
                  numberOfLines={4}
                  style={
                    Array [
                      Object {
                        "color": "#ffffff",
                        "fontFamily": "Poppins-SemiBold",
                        "fontSize": 14,
                        "zIndex": 5,
                      },
                    ]
                  }
                />
              </View>
            </View>
          </View>
        </View>
      </View>
    </RCTScrollView>
  </RCTSafeAreaView>
</View>
`;
